### PR TITLE
[Merged by Bors] - ET-3560 rank by similarity & category

### PR DESCRIPTION
**Pull request by janpetschexain**
**Reference**

- [ET-3560]
- requires #693

**Summary**

- add sorting of documents by categories next to the sorting of documents by scores/similarities and combine the two rankings

[ET-3560]: https://xainag.atlassian.net/browse/ET-3560?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/web-api/src/personalization/routes.rs [hunk 1]
```diff
@@ -216,13 +216,56 @@ async fn personalized_documents(
...
+        a.cmp(&b).reverse()
+    });
+    let documents_by_categories = documents_by_categories
+        .into_iter()
+        .enumerate()
+        .map(|(rank, document)| (document.id.clone(), rank))
+        .collect::<HashMap<_, _>>();
 
+    let weight = state.config.personalization.rank_weight;
+    all_documents.sort_unstable_by(|a, b| {
```
[the remark this answers is not in the document]
Comment (janpetschexain) [reply discovery_engine_core/web-api/src/personalization/routes.rs:264]: essentially a weighted sum of the two ranks. we have the documents ranked by user interests (via the cosine similarity with the cois) and the documents ranked by the category weights. then we compute the weighted sum `document_rank_by_interest * weight + document_rank_by_category * (1-weight)` and order the documents according to this new score. a tie is either broken according to the higher `weight`/`1-weight` or randomly if both are the same.

File: discovery_engine_core/web-api/src/personalization/routes.rs [hunk 2]
```diff
@@ -235,6 +235,47 @@ async fn personalized_documents(
...
+        })
+        .collect_vec();
+    documents_by_categories.sort_unstable_by(|(_, a), (_, b)| a.cmp(b).reverse());
+    let documents_by_categories = documents_by_categories
+        .into_iter()
+        .enumerate()
+        .map(|(rank, (document_id, _))| (document_id, rank))
+        .collect::<HashMap<_, _>>();
+
+    let weight = state.config.personalization.rank_weight;
```
Comment (acrrd) [inline discovery_engine_core/web-api/src/personalization/routes.rs:263]: :art:
I think this name is too generic. `interests_cat_bias` ? or any other name that is a bit more explicit about what this is.

**Review by acrrd — APPROVED** [review]

File: discovery_engine_core/web-api/src/personalization/routes.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (PanicButtonPressed) [reply discovery_engine_core/web-api/src/personalization/routes.rs:262]: Ok thanks for the explainer. Totally makes sense. Was just a bit confused by the names weighted_a and weighted_b

**Review by PanicButtonPressed — APPROVED** [review]
Well done

**Comment by janpetschexain** [comment]
bors merge

**Pull request closed without merging**